### PR TITLE
Quest: Sleeping Giants

### PR DESCRIPTION
**Pull request by claudiodekker**
This PR adds a mostly-complete implementation of the Sleeping Giants quest. 

- It still needs some verification, and probably some 'patching', as Runelite hasn't been updated yet to 'contain' the right NPC and Item ID's, so for now they're hard-coded integers but they work.
- For the same reason, the quest id isn't set properly yet. I hope someone else can fix this when Runelite is updated!

As for the main concerns of things I might've screwed up & cannot verify:
- A big part of the quest is based around a single varbit that increments/changes value, and I might've accidentally screwed up some of the checks (There's a TODO for it in the code)
- At a certain point you need to set a mould, and I logged out/logged back in during this causing the step to be set. It'd be appreciated if someone can verify whether this step & it's continuance from there is working properly.
- At a certain point, you need to take a "preform" item from a mould (game object). I kinda botched this step, so am not 100% sure what triggers the change between the 'interacting with object' vs 'starting the learning minigame'. I'm assuming the change happens due to the player picking up the preform, but it might as well be because of a varbit change. I've retroactively implemented a post-quest variant of the perform item on this assumption, but am not sure if this works.

In addition, as you can tell from the large amount of changes, **I've moved some of the WidgetRequirement types to a 'widget' subfolder, just like the others**.

### CODE REVIEW

File: src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+	public void setupConditions()
+	{
+		grindstoneFixed = new VarbitRequirement(13905, 2, Operation.EQUAL);
+		polishingStoneFixed = new VarbitRequirement(13906, 2, Operation.EQUAL);
+		hammerFixed = new VarbitRequirement(13904, 2, Operation.EQUAL);
+
+		commissionReceived = new VarbitRequirement(13903, 10, Operation.EQUAL);
+		crateSearched = new VarbitRequirement(13903, 25, Operation.LESS_EQUAL);
+		shouldSetMould = new VarbitRequirement(13903, 30, Operation.LESS_EQUAL);
```
Comment (xavierbroadhead) [inline src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java:189]: Why are these operations using the LESS_EQUAL operator? I changed them to EQUAL locally and they worked fine.

File: src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+	}
+
+	public void setupConditions()
+	{
+		grindstoneFixed = new VarbitRequirement(13905, 2, Operation.EQUAL);
+		polishingStoneFixed = new VarbitRequirement(13906, 2, Operation.EQUAL);
+		hammerFixed = new VarbitRequirement(13904, 2, Operation.EQUAL);
+
+		commissionReceived = new VarbitRequirement(13903, 10, Operation.EQUAL);
+		crateSearched = new VarbitRequirement(13903, 25, Operation.LESS_EQUAL);
```
Comment (xavierbroadhead) [inline src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java:188]:
```suggestion
		crateSearched = new VarbitRequirement(13903, 15, Operation.EQUAL);
```

Changing the Varbit value here should fix this step.

File: src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+		speakToKovacForCommission.addStep(new Conditions(commissionReceived), searchCrate);
+		speakToKovacForCommission.addStep(new Conditions(crateSearched), fillCrucible);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noForteSelected, forteTabOpen), selectForte);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noForteSelected), selectForteTab);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noBladeSelected, bladeTabOpen), selectBlade);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noBladeSelected), selectBladesTab);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noTipSelected, tipTabOpen), selectTip);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noTipSelected), selectBlade);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould), setMould);
+		speakToKovacForCommission.addStep(new Conditions(shouldSetMould), interactWithMould);
```
Comment (xavierbroadhead) [inline src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java:112]: There needs to be an additional step before this to speak to Kovac again before setting the mould.

File: src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+		ConditionalStep speakToKovacContinueAfterRepairs = new ConditionalStep(this, speakToKovacAfterRepairs);
+		speakToKovacContinueAfterRepairs.addStep(new Conditions(LogicType.NOR, inGiantsFoundry), enterFoundry);
+		steps.put(20, speakToKovacContinueAfterRepairs);
+
+		// TODO: This needs validation, as it's a very complex varbit-base step. It is very possible that the varbits are not checked properly.
+		ConditionalStep speakToKovacForCommission = new ConditionalStep(this, speakToKovacContinue);
+		speakToKovacForCommission.addStep(new Conditions(LogicType.NOR, inGiantsFoundry), enterFoundry);
+		speakToKovacForCommission.addStep(new Conditions(commissionReceived), searchCrate);
+		speakToKovacForCommission.addStep(new Conditions(crateSearched), fillCrucible);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noForteSelected, forteTabOpen), selectForte);
```
Comment (xavierbroadhead) [inline src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java:105]: These steps are out of order - moving them to below where the interactWithMould step is added would be a bit more readable.

File: src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noForteSelected), selectForteTab);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noBladeSelected, bladeTabOpen), selectBlade);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noBladeSelected), selectBladesTab);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noTipSelected, tipTabOpen), selectTip);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould, noTipSelected), selectBlade);
+		speakToKovacForCommission.addStep(new Conditions(selectingMould), setMould);
+		speakToKovacForCommission.addStep(new Conditions(shouldSetMould), interactWithMould);
+		speakToKovacForCommission.addStep(new Conditions(mouldSet), speakToKovac);
```
Comment (xavierbroadhead) [inline src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java:113]: Reusing the speakToKovac step brings the plugin back to a previously completed step, from "Creating the sword" sub-category to the "Starting out" category. The NPC ID for the speakToKovac step also points to the previous model found at the start of the quest, not the Kovac NPC in the Foundry itself.

File: src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+		enterFoundry = new ObjectStep(this, 44635 /* Cave */, new WorldPoint(3361, 3150, 0), "Enter the Giants' Foundry.");
+
+		fixPolishingStone = new ObjectStep(this, 44775 /* Broken polishing wheel */, new WorldPoint(3363, 11485, 0), "Fix the Broken polishing wheel", twoOakLogs, fiveNails, hammer);
+		fixPolishingStone.addDialogStep("Yes.");
+		fixGrindstone = new ObjectStep(this, 44774 /* Broken grindstone */, new WorldPoint(3362, 11492, 0), "Fix the Broken grindstone", chisel);
+		fixGrindstone.addDialogStep("Yes.");
+		fixHammer = new ObjectStep(this, 44773 /* Broken trip hammer */, new WorldPoint(3364, 11497, 0), "Fix the Broken trip hammer", oneOakLog, fiveNails, hammer);
+		fixHammer.addDialogStep("Yes.");
+
+		speakToKovacAfterRepairs = new NpcStep(this, 11468 /* Kovac */, "Speak to Kovac to continue.");
```
Comment (xavierbroadhead) [inline src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java:223]: This NPC ID is incorrect. 
```suggestion
		speakToKovacAfterRepairs = new NpcStep(this, 11470 /* Kovac */, "Speak to Kovac to continue.");
```

**Review by xavierbroadhead — CHANGES_REQUESTED** [review]
It looks mostly fine to me. There are a few NPC ID and Varbit value issues, as well as some code style issues that should be addressed before this is ready for further testing.

File: src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java (hunk 4 above)
Comment (Zoinkwiz) [reply src/main/java/com/questhelper/quests/sleepinggiants/SleepingGiants.java:142]: Should work correctly now!

**Review by Zoinkwiz — APPROVED** [review]
Looks good to me now, added a few fixes and finished the final section. I'll leave it unmerged for now until we're clear to release it though.

**Pull request merged**